TWTValidation/JSON Validator/Processors/TWTJSONObjectValidatorGenerator.m

```diff
@@ -0,0 +1,410 @@
+//
+//  TWTJSONObjectValidatorGenerator.m
+//  TWTValidation
+//
+//  Created by Jill Cohen on 1/14/15.
+//  Copyright (c) 2015 Two Toasters, LLC.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+#import <TWTValidation/TWTJSONObjectValidatorGenerator.h>
+
+#import <TWTValidation/TWTValidation.h>
+
+#import <TWTValidation/TWTJSONSchemaASTCommon.h>
+#import <TWTValidation/TWTJSONSchemaParser.h>
+#import <TWTValidation/TWTJSONSchemaArrayValidator.h>
+#import <TWTValidation/TWTJSONSchemaObjectValidator.h>
+
+
+@interface TWTJSONObjectValidatorGenerator ()
+
+@property (nonatomic, strong, readonly) NSMutableArray *objectStack;
+
+@end
+
+
+@implementation TWTJSONObjectValidatorGenerator
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self) {
+        _objectStack = [[NSMutableArray alloc] init];
+    }
+    return self;
+}
+
+
+- (TWTJSONObjectValidator *)validatorFromJSONSchema:(NSDictionary *)schema error:(NSError *__autoreleasing *)outError warnings:(NSArray *__autoreleasing *)outWarnings
+{
+    [self.objectStack removeAllObjects];
+
+    TWTJSONSchemaParser *parser = [[TWTJSONSchemaParser alloc] initWithJSONSchema:schema];
+    NSError *parsingError = nil;
+    TWTJSONSchemaTopLevelASTNode *topLevelNode = [parser parseWithError:&parsingError warnings:outWarnings];
+    if (parsingError) {
```

Comment:
As a general pattern, you shouldn’t check the `parsingError` parameter unless the return value of a method is `nil` or `NO`. Cocoa usually doesn’t guarantee that it won’t modify your error pointer even when an error didn’t occur. It does guarantee that a method will return `nil` or `NO`. This works out well for you, because you can eliminate `parsingError` entirely and just pass `outError`. Then if `topLevelNode` is `nil`, just return `nil`.